create task to write merged forced_src tables to parquet

timothydmorton opened the pull request:
This new CmdLineTask builds on MergeSourcesTask to merge
`deepCoadd_forced_src`, `deepCoadd_meas`, and `deepCoadd_ref`
tables from multiple filters into single
joined parquet tables, one per patch.  The output dataset is
`deepCoadd_object`, and gets loaded as a pandas dataframe
with a multi-level column index, such that you can retrieve
all the columns of a single table with, e.g., df['meas']['HSC-R'].

timothydmorton committed:
create task to write merged forced_src tables to parquet

This new CmdLineTask builds on MergeSourcesTask to merge
deepCoadd_forced_src, deepCoadd_meas, and deepCoadd_ref
tables from multiple filters into single
joined parquet tables, one per patch.  The output dataset is
deepCoadd_object, and gets loaded as a pandas dataframe
with a multi-level column index, such that you can retrieve
all the columns of a single table with, e.g., df['meas']['HSC-R'].

timothydmorton commented:
Sorry about all the micro-commits... I thought I had squashed all that into 23fed10...

parejkoj commented on python/lsst/qa/explorer/table.py:
```diff
@@ -0,0 +1,33 @@
+
```
File level docstring and copyright/license.

timothydmorton replied:
This file will be replaced in DM-13876

parejkoj commented on python/lsst/qa/explorer/table.py:
```diff
@@ -0,0 +1,33 @@
+    Parameters
+    ----------
+    df : pandas.DataFrame
+
```
Document the `engine` kwarg.

parejkoj commented on python/lsst/qa/explorer/table.py:
```diff
@@ -0,0 +1,33 @@
+    def readFits(cls, filename):
+        """Read parquet file (not FITS) into pandas DataFrame.
+        """
+        return pd.read_parquet(filename, engine='pyarrow')
```
Need newline at end of file. Are you running flake8?

timothydmorton replied:
Nope, I'm not.  Perhaps I should be?

parejkoj commented on python/lsst/qa/explorer/table.py:
```diff
@@ -0,0 +1,33 @@
+    def readFits(cls, filename):
+        """Read parquet file (not FITS) into pandas DataFrame.
+        """
+        return pd.read_parquet(filename, engine='pyarrow')
```
Did mean to use `self.engine` here, or should this always use `pyarrow`?

timothydmorton replied:
I was going back and forth between whether to have the engine be configurable or not, and I sort of went halfway... there was a reason (I think because I wanted to test writing with both engines but then there was a bug in fastparquet that made reading really slow)

parejkoj commented on python/lsst/qa/explorer/writeParquet.py:
```diff
@@ -0,0 +1,107 @@
+from lsst.daf.persistence.butler import Butler
```
File level docstring and copyright/license.

timothydmorton replied:
Did the docstring; I'll get to copyright in qa_explorer at some point.

parejkoj commented on python/lsst/qa/explorer/table.py:
```diff
@@ -0,0 +1,33 @@
+        """
+        if self.engine=='pyarrow':
+            table = pa.Table.from_pandas(self.df)
+            pq.write_table(table, filename, compression='none')
```
Would it be worth making compression an option?

timothydmorton replied:
Also punting this to DM-13876

parejkoj commented on python/lsst/qa/explorer/writeParquet.py:
```diff
@@ -0,0 +1,107 @@
+class WriteObjectTableConfig(MergeSourcesConfig):
+    priorityList = ListField(dtype=str, default=['HSC-G', 'HSC-R', 'HSC-I', 'HSC-Z', 'HSC-Y'],
+                             doc="Priority-ordered list of bands for the merge.")    
+    engine = Field(dtype=str, default="pyarrow", doc="Parquet engine for writing (pyarrow or fastparquet)")
```
If those are the only options, make it a `ChoiceField`.

timothydmorton replied:
They aren't options; this is supposed to be a list:  https://github.com/lsst/pipe_tasks/blob/cf273cdbbb95f59b14fd3f4e6cddfab8dfd9637f/python/lsst/pipe/tasks/multiBand.py#L389

parejkoj commented on python/lsst/qa/explorer/writeParquet.py:
```diff
@@ -0,0 +1,107 @@
+    outputDataset = 'obj'
+
+    # Hack, since we're not using this...
+    getSchemaCatalogs = lambda x : {}
```
What's the hack? If you're overriding something that you want to do nothing, you should do something like this:

```
def getSchemaCatalogs(self, x):
    """Overriding the default to do nothing."""
    pass
```

timothydmorton replied:
The hack is that this function needs to return a dictionary.  I can write a proper method rather than a lambda, I suppose.

parejkoj replied:
Yes, please write a function.

timothydmorton replied:
(done already)

parejkoj commented on python/lsst/qa/explorer/writeParquet.py:
```diff
@@ -0,0 +1,107 @@
+    def _makeArgumentParser(cls):
+        """Create a suitable ArgumentParser.
+
+        We will use the ArgumentParser to get a provide a list of data
```
redundancy? "...to get a provide a list..."

timothydmorton replied:
Ha! I just copied from https://github.com/lsst/pipe_tasks/blob/cf273cdbbb95f59b14fd3f4e6cddfab8dfd9637f/python/lsst/pipe/tasks/multiBand.py#L432 without reading carefully!

parejkoj commented on python/lsst/qa/explorer/writeParquet.py:
```diff
@@ -0,0 +1,107 @@
+    ConfigClass = WriteObjectTableConfig
+
+    inputDatasets = ('forced_src', 'meas', 'ref')
+    outputDataset = 'obj'
```
Should document these two attributes, and/or configure them as part of `__init__`, particularly given how `self.inputDatasets` is used below.

parejkoj commented on python/lsst/qa/explorer/writeParquet.py:
```diff
@@ -0,0 +1,107 @@
+    def readCatalog(self, patchRef):
+        """Read input catalogs
+
+        Read all the input datasets provided by the 'inputDatasets'
```
I think for "provided" here you mean "given"?

parejkoj commented on python/lsst/qa/explorer/writeParquet.py:
```diff
@@ -0,0 +1,107 @@
+        name
+        """
+        filterName = patchRef.dataId["filter"]
+        d = {}
```
Please don't use single character variable names for anything other than iterators. I think this is the `catalogDict`?

parejkoj commented on python/lsst/qa/explorer/writeParquet.py:
```diff
@@ -0,0 +1,107 @@
+
+        Returns
+        -------
+        catalog : lsst.qa.explorer.table.ParquetTable
```
backticks around type.

parejkoj commented on python/lsst/qa/explorer/writeParquet.py:
```diff
@@ -0,0 +1,107 @@
+
+        Parameters
+        ----------
+        catalogs : dict
```
backticks around type.

parejkoj commented on python/lsst/qa/explorer/writeParquet.py:
```diff
@@ -0,0 +1,107 @@
+                                                       names=('dataset', 'filter', 'column'))
+                dfs.append(df)
+
+        catalog = functools.reduce(lambda d1,d2 : d1.join(d2), dfs)
```
Right, that's how you do n-way table join... I was just thinking about this with `astropy.table.join`. Thanks!

parejkoj commented on python/lsst/qa/explorer/writeParquet.py:
```diff
@@ -0,0 +1,107 @@
+                             doc="Priority-ordered list of bands for the merge.")    
+    engine = Field(dtype=str, default="pyarrow", doc="Parquet engine for writing (pyarrow or fastparquet)")
+
+class WriteObjectTableTask(MergeSourcesTask):
```
Should this be called `WriteOjbectTableAsParquetTask` or `WriteParquetTableTask`, just to be more explicit?

timothydmorton replied:
I'm sticking with `WriteObjectTableTask`.  Users shouldn't care that it's parquet if we do our job right.

parejkoj reviewed (changes requested):
The `readFits/writeFits` faking you do here worries me, but I guess adding parquet support to the current butler is probably a lot more work. It does seem a bit abusive of the butler's trust...

It would probably be good to have a some round-tripping tests for this: start with a few minimal catalogs, send them to parquet, check that you get back what you sent.

Tests would also make it easier for me to see an example if I wanted to use it myself.

timothydmorton commented:
OK, I've just squashed everything into a single commit here, which may not have been the greatest idea, as it obscures the corrections in response to the review.  But I think I've addressed everything!
